shell/platform/android/io/flutter/embedding/engine/android/AndroidKeyProcessor.java

```diff
@@ -0,0 +1,87 @@
+package io.flutter.embedding.engine.android;
+
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+import android.view.KeyCharacterMap;
+import android.view.KeyEvent;
+
+import io.flutter.embedding.engine.systemchannels.KeyEventChannel;
+
+public class AndroidKeyProcessor {
+  private final KeyEventChannel keyEventChannel;
+  private int combiningCharacter;
+
+  public AndroidKeyProcessor(@NonNull KeyEventChannel keyEventChannel) {
+    this.keyEventChannel = keyEventChannel;
+  }
+
+  public void onKeyUp(@NonNull KeyEvent keyEvent) {
+    Character complexCharacter = applyCombiningCharacterToBaseCharacter(keyEvent.getUnicodeChar());
+    keyEventChannel.keyUp(
+        new KeyEventChannel.FlutterKeyEvent(keyEvent, complexCharacter)
+    );
+  }
+
+  public void onKeyDown(@NonNull KeyEvent keyEvent) {
+    Character complexCharacter = applyCombiningCharacterToBaseCharacter(keyEvent.getUnicodeChar());
+    keyEventChannel.keyDown(
+        new KeyEventChannel.FlutterKeyEvent(keyEvent, complexCharacter)
+    );
+  }
+
+  /**
+   * Applies the given Unicode character in {@code newCharacterCodePoint} to a previously
+   * entered Unicode combining character and returns the combination of these characters
+   * if a combination exists.
+   * <p>
+   * This method mutates {@link #combiningCharacter} over time to combine characters.
+   * <p>
+   * One of the following things happens in this method:
+   * <ul>
+   *   <li>If no previous {@link #combiningCharacter} exists and the {@code newCharacterCodePoint}
+   *   is not a combining character, then nothing happens and null is returned.</li>
+   *   <li>If no previous {@link #combiningCharacter} exists and the {@code newCharacterCodePoint}
+   *   is a combining character, then {@code newCharacterCodePoint} is saved as the
+   *   {@link #combiningCharacter} and null is returned.</li>
+   *   <li>If a previous {@link #combiningCharacter} exists and the {@code newCharacterCodePoint}
+   *   is also a combining character, then the {@code newCharacterCodePoint} is combined with
+   *   the existing {@link #combiningCharacter} and null is returned.</li>
+   *   <li>If a previous {@link #combiningCharacter} exists and the {@code newCharacterCodePoint}
+   *    is not a combining character, then the {@link #combiningCharacter} is applied to the
+   *    regular {@code newCharacterCodePoint} and the resulting complex character is returned. The
+   *    {@link #combiningCharacter} is cleared.</li>
+   * </ul>
+   * <p>
+   * The following reference explains the concept of a "combining character":
+   * https://en.wikipedia.org/wiki/Combining_character
+   */
+  @Nullable
+  private Character applyCombiningCharacterToBaseCharacter(int newCharacterCodePoint) {
+    if (newCharacterCodePoint == 0) {
+      return null;
+    }
+
+    Character complexCharacter = null;
```

Comment:
I think it actually should be this, since if the new character is just a regular character, and there wasn't a preceding combining character (the common case) we want to return it.
```suggestion
    Character complexCharacter = (char) newCharacterCodePoint;
```